tests/sentry/auth/test_helper.py

```diff
@@ -77,25 +120,147 @@ def test_associated_existing_member_invite_request(self):
+        assert mock_auth.get_login_redirect.called_with(self.request)
+
+        persisted_identity = AuthIdentity.objects.get(ident=auth_identity.ident)
+        assert persisted_identity.data == self.identity["data"]
+
+        persisted_om = OrganizationMember.objects.get(user=user, organization=self.organization)
+        assert getattr(persisted_om.flags, "sso:linked")
+        assert not getattr(persisted_om.flags, "sso:invalid")
+
+        login_request, login_user = mock_auth.login.call_args.args
+        assert login_request == self.request
+        assert login_user == user
+
+
+class HandleAttachIdentityTest(AuthIdentityHandlerTest):
+    def _handle_attach_identity(self):
+        return handle_attach_identity(
+            self.auth_provider, self.request, self.organization, self.provider_obj, self.identity
+        )
+
+    @mock.patch("sentry.auth.helper.messages")
+    def test_new_identity(self, mock_messages):
+        self.set_up_user()
+
+        auth_identity = self._handle_attach_identity()
+        assert auth_identity.ident == self.identity["id"]
+        assert auth_identity.data == self.identity["data"]
+
+        assert AuditLogEntry.objects.filter(
+            organization=self.organization,
+            target_object=auth_identity.id,
+            event=AuditLogEntryEvent.SSO_IDENTITY_LINK,
+            data=auth_identity.get_audit_log_data(),
+        ).exists()
+
+        assert mock_messages.add_message.called_with(
+            self.request, messages.SUCCESS, OK_LINK_IDENTITY
+        )
+
+    @mock.patch("sentry.auth.helper.messages")
+    def test_existing_identity(self, mock_messages):
+        user, existing_identity = self.set_up_user_identity()
+
+        returned_identity = self._handle_attach_identity()
+        assert returned_identity == existing_identity
+        assert not mock_messages.add_message.called
+
+    def test_wipe_other_identity(self):
+        request_user, existing_identity = self.set_up_user_identity()
+        other_profile = self.create_user()
+
+        # The user logs in with credentials from this other identity
+        AuthIdentity.objects.create(
+            user=other_profile, auth_provider=self.auth_provider, ident=self.identity["id"]
+        )
+        OrganizationMember.objects.create(user=other_profile, organization=self.organization)
+
+        returned_identity = self._handle_attach_identity()
+        assert returned_identity.ident == self.identity["id"]
+        assert returned_identity.data == self.identity["data"]
+
+        assert not AuthIdentity.objects.filter(id=existing_identity.id).exists()
+
+        persisted_om = OrganizationMember.objects.get(
+            user=other_profile, organization=self.organization
+        )
+        assert not getattr(persisted_om.flags, "sso:linked")
+        assert getattr(persisted_om.flags, "sso:invalid")
+
+
+class HandleUnknownIdentityTest(AuthIdentityHandlerTest):
```

Comment:
only looking at current branch to compare unsure if your new one reduces complexity, but i suppose we could add a lot more tests for this function -- also not sure how much functionality is dupe tested in the integration tests so i suppose this may be a good middleground.

Reply:
We definitely could. There are a lot of `if op == "..."` branches that need coverage.

Because this was done to support #26571 and that PR doesn't do anything too risky within the guts of `handle_unknown_identity`, adding really detailed coverage there seemed like a high cost-to-benefit ratio, so I pretty much skipped it here. It's worth a TODO comment, however.